pkg/provisioner/crdprovisioner.go

```diff
@@ -375,72 +380,80 @@ func (c *CrdProvisioner) PublishVolume(
-			return true, nil
-		}
-		if azVolumeAttachmentInstance.Status.Error != nil {
-			return false, util.ErrorFromAzError(azVolumeAttachmentInstance.Status.Error)
-		}
-		return false, nil
-	})
+	// return error if volume is not found
+	azVolume, err := volumeLister.Get(volumeName)
+	if errors.IsNotFound(err) {
+		return nil, status.Errorf(codes.NotFound, "volume (%s) does not exist", volumeName)
+	}
 
-	if err != nil {
-		return nil, err
+	// return error if node is not found
+	if _, err := nodeLister.Get(nodeID); errors.IsNotFound(err) {
+		return nil, status.Errorf(codes.NotFound, "node (%s) does not exist", nodeID)
 	}
-	defer waiter.Close()
 
 	if volumeContext == nil {
 		volumeContext = map[string]string{}
 	}
-	azVolumeAttachmentInstance, err := lister.Get(attachmentName)
-	if err == nil {
-		// if CRI is scheduled for deletion, retry until operations are complete
-		if azVolumeAttachmentInstance.DeletionTimestamp != nil {
-			return nil, status.Errorf(codes.Aborted, "need to wait until attachment is fully deleted for volume (%s) and node (%s) before attaching", volumeName, nodeID)
+
+	publishContext := map[string]string{}
+
+	// preemptively detach any replica volume if necessary
+	if azVolume.Spec.MaxMountReplicaCount > 0 {
+		// get undemoted AzVolumeAttachments for volume == volumeName and node != nodeID
+		volumeLabel, _ := azureutils.CreateLabelRequirements(consts.VolumeNameLabel, selection.Equals, volumeName)
+		nodeLabel, _ := azureutils.CreateLabelRequirements(consts.NodeNameLabel, selection.NotEquals, nodeID)
+		roleChangeLabel, _ := azureutils.CreateLabelRequirements(consts.RoleChangeLabel, selection.NotEquals, consts.Demoted)
+		azVASelector := labels.NewSelector().Add(*volumeLabel, *nodeLabel, *roleChangeLabel)
+		undemotedAttachments, err := azVALister.List(azVASelector)
+		if err != nil && !errors.IsNotFound(err) {
+			return nil, status.Errorf(codes.Internal, "failed to list undemoted AzVolumeAttachments with volume (%s) not attached to node (%s): %v", volumeName, nodeID, err)
 		}
-		// If there already exists a primary attachment with populated responseObject return
-		if azVolumeAttachmentInstance.Status.Detail != nil && azVolumeAttachmentInstance.Status.Detail.PublishContext != nil && azVolumeAttachmentInstance.Status.Detail.Role == diskv1alpha2.PrimaryRole {
-			return azVolumeAttachmentInstance.Status.Detail.PublishContext, nil
-			// if attachment is pending return error to prevent redundant attachment request
-		} else if azVolumeAttachmentInstance.Status.State == diskv1alpha2.Attaching {
-			return nil, status.Errorf(codes.Aborted, "attachment still in process for volume (%s) and node (%s)", volumeName, nodeID)
+
+		// get demoted AzVolumeAttachments
+		roleChangeLabel, _ = azureutils.CreateLabelRequirements(consts.RoleChangeLabel, selection.Equals, consts.Demoted)
+		azVASelector = labels.NewSelector().Add(*volumeLabel, *nodeLabel, *roleChangeLabel)
+		demotedAttachments, err := azVALister.List(azVASelector)
```

Comment:
Can we instead call azVALister.List only once and then loop over them to segregate the demoted and the non demoted attachments?

Reply:
We can, but we would need to sort them based on whether the attachment is demoted or not. I wonder what would be more ideal here: listing them twice from cached client vs listing once and sorting them.

Reply:
let
  Ag be the global attachment count
  Av be the attachment count for a volume
where
  Av <= Ag

A single operations is O(Ag)
A  sort is O(Av log Av)

Two lists seem less efficient that a single list followed by sort.

Reply:
I wouldn't block this PR because of this, but it would be good to address in the future.

Reply:
I had imagined that the list operations using label selectors would be using some sort of mapping scheme to avoid iterating through all object instances but if not, I agree that we should restrain from calling multiple list operations when possible.

Reply:
Nope, it's a linear search/ See [`cache.ListAll`](https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/c7c00ad166e22433d40f5ddcb156f51911890939/vendor/k8s.io/client-go/tools/cache/listers.go#L34) which is called by the listers' [`List`](https://github.com/abhisheksinghbaghel/azuredisk-csi-driver/blob/b257560f19797e8782077a357c4d85086e5bf94c/pkg/apis/client/listers/azuredisk/v1alpha2/azvolumeattachment.go#L50) method.